website/content/docs/concepts/delivery-semantics.md

```diff
@@ -0,0 +1,62 @@
+
+    If the input to the topology is, for example, a messaging system that cannot ensure stream consistency, then effectively-once semantics cannot be applied, as the state "rewind" may return differing results. To put it somewhat differently, Heron can only provide delivery semantics as stringent as its data input sources can themselves provide.
+
+### Exactly-once semantics?
+
+There has been a lot of discussion recently surrounding so-called "exactly-once" processing semantics. We'll avoid this term in the Heron documentation because we feel that it's misleading. "Exactly-once" semantics would mean that no processing step is ever performed more than once---and thus that no processing step is ever retried.
+
+It's important to always keep in mind that *no system* can provide exactly-once semantics in the face of failures (as [this article](http://bravenewgeek.com/you-cannot-have-exactly-once-delivery) argues). But that's okay because they don't really need to; the truly important thing is that a stream processing system be able to recover from failures by "rewinding" state to a previous, pre-failure point and to re-attempt to apply processing logic. We use the tern **effectively once**, following [Victor Klang](https://twitter.com/viktorklang/status/789036133434978304), for this style of semantics.
+
+Heron *can* provide effectively-once guarantees if a topology meets the conditions [outlined above](#requirements-for-effectively-once), but it cannot provide "exactly-once" semantics.
+
+## Stateful topologies
+
+The Heron topologies that you create can be either stateful or non stateful.
+
+* In **stateful topologies**, each component must implement an interface that requires it to store its state every time it processes a tuple (both spouts *and* bolts must do so).
+* In **non-stateful topologies**, there is no requirement that any processing components store a state snapshot. Non-stateful topologies can provide at-most-once or at-least-once semantics, but never effectively-once semantics.
+
+> Heron currently supports two state managers: [ZooKeeper](../../operators/deployment/statemanagers/zookeeper) and the [local filesystem](../../operators/deployment/statemanagers/localfs), although others are currently under development.
+
+Stateful topologies, in turn, are of two types:
+
+* **Idempotent** stateful topologies are stateful topologies in which applying the processing graph to an input more than once, it will continue to return the same result. A basic example is multiplying a number by 0. The first time you do so, the number will change (always to 0), but if you apply that transformation again and again, it will not change.
+
+    For topologies to provide effectively-once semantics, they need to transform tuple inputs idempotently as well. If they don't, and applying the topology's processing graph multiple times yields different results, then effectively-once semantics *cannot* be achieved.
```

Comment:
if topology is idempotent, at-least once can achieve the same result ? no need to enable `effectively once`?

Reply:
I would reword this as
they need to ensure that any state changes caused by execution of the tuple is captured in the heron state variable thats passed at startup time. This way Heron can keep track of state of the processing logic and rewind state when failures occur.

Reply:
@huijunw As far as I know, idempotency and effectively once correspond to different concerns. A topology can be idempotent and process tuples more than once, which is a problem if you have effectively once requirements.